### DISCUSSION TITLE
Converting PNG to raw RGB565 Format

### DISCUSSION THREAD

**Discussion by crim491** [ImageMagick/ImageMagick#2787] [Help] [answered]
I'm working on a embedded system that has uses the ili9341 display. This display has an image format of RGB656 in 12bit mode. I'm trying to take a source PNG file and convert it to a raw RGB565 format. My reasoning for doing it is to determine the methodology required to convert other source files to a PNG file with this format. The best I was able to come up with is: 
"magick convert blu.png -size 151x320 -alpha remove -depth 1 -ordered-dither threshold,32,64,32 RGBA:bluerbg.raw". 

I'm expecting an output in the raw file  to be filled with "0x001F" but what I get is "0x33". I am not entirely sure how to interpret this output other than it is not what I want. "blu.png" is a 151x320 image that has a solid blue fill.

I would initially convert "blu.png" to "bluc.png" using "convert blu.png -colorspace rgb -depth1 -alpha remove -ordered-dither threshold,32,64,32 bluc.png"

Thoughts? 

Thanks.

**Comment by fmw42** [comment]
I am not an expert on this. But you could threshold or quantize each channel separately. Another thought would be to create a BMP RGB565 file and then possibly convert that to PNG. See https://imagemagick.org/script/defines.php -define bmp:subtype=RGB565. But I am not sure PNG supports such a format. But again I am not an expert on PNG formats.

**Comment by t246246** [comment, marked as answer]
If you want raw file, see snibgo's answer here: https://github.com/ImageMagick/ImageMagick/discussions/2559 

EDIT:
After some trial, RGB: (or RGBA:) prefix saves the result RGB888 (or 8888).
There is GRAY: prefix but this does not help in this case.
Also there is RGB565: prefix, but only reading is suppoted.
```
% magick -list format | grep -e '\(RGB\|GRAY\)'
     GRAY* GRAY      rw+   Raw gray samples
    GRAYA* GRAY      rw+   Raw gray and alpha samples
      HDR* HDR       rw+   Radiance RGBE image format
    PNG24* PNG       rw-   opaque or binary transparent 24-bit RGB (zlib 1.2.11)
    PNG32* PNG       rw-   opaque or transparent 32-bit RGBA
    PNG48* PNG       rw-   opaque or binary transparent 48-bit RGB
    PNG64* PNG       rw-   opaque or transparent 64-bit RGBA
      RGB* RGB       rw+   Raw red, green, and blue samples
   RGB565* RGB       r--   Raw red, green, blue samples in 565 format
     RGBA* RGB       rw+   Raw red, green, blue, and alpha samples
     RGBO* RGB       rw+   Raw red, green, blue, and opacity samples
      SGI* SGI       rw+   Irix RGB image
```

So, it is not a good way but create bmp565 first using -define bmp:subtype=RGB565.(Note that if png is palette type, this may not work. See https://www.imagemagick.org/discourse-server/viewtopic.php?t=33691.)
Then remove BMP header and padding.
Do something like this:
```
#!/bin/sh

w=`magick identify -ping -format '%w' $1`
magick $1 -flip -type truecolor -define bmp:subtype=rgb565 BMP:- |
dd bs=1 skip=138 2>/dev/null |
perl -e "my(\$w) = $w;" \
     -e 'my($r) = $w+(4-$w%4)%4;
         print substr($buf, 0, $w*2) while(read(STDIN, $buf, $r*2));'
```

Or use ffmpeg:
% ffmpeg -vcodec png -i blu.png -vcodec rawvideo -f rawvideo -pix_fmt rgb565 blu.raw

It looks the result differs a little between ffmpeg and `-define bmp:subtype=RGB565`

**Comment by snibgo** [comment]
> ... RGB656 in 12bit mode ... raw RGB565 format ...

Huh? Perhaps you want RGB565, so you effectively have one channel of 16 bits, where the first 5 bits are for red, the next 6 are for green, and the final 5 are for green. Your output should have just these 16-bit numbers, with no header. Is that what you want?

We can do that with some fiddly arithmetic. Windows BAT syntax:
```
magick ^
  in.png ^
  -channel R -evaluate Multiply %%[fx:31/QuantumRange] -evaluate LeftShift 11 ^
  -channel G -evaluate Multiply %%[fx:63/QuantumRange] -evaluate LeftShift 5 ^
  -channel B -evaluate Multiply %%[fx:31/QuantumRange] ^
  +channel ^
  -separate ^
  -evaluate-sequence Add ^
  gray:out.bin
```

**Reply by crim491** [reply]
I know nothing about image processing/representation/manipulation and as a result much of this is very foreign to me. What I do know is that the LCD controller has a pixel format of [rrrrrggggggbbbbb]. Doing fiddly arithmetic makes me extremely nervous because I wouldn't know if the operation plays through correctly.

**Reply by BurgessShale** [reply]
I ran this .bat file on my .png image, then opened the resulting .bin in a hex editor-- and every nonzero pixel was 0xFF. So it doesn't seem to thunk properly.

**Comment by snibgo** [comment]
I am guessing at what you need. If you can provide a sample image that works for you, that would reduce my need to guess.

>  ... I wouldn't know if the operation plays through correctly.

You can test that.

For example, when I use my usual image toes.png which is 267x233 pixels, I can read my binary file and compare it to test.png:
```
%IMG7%magick compare ^
  -metric RMSE ^
  toes.png ^
  -size 267x233 RGB565:x.bin ^
  NULL:

1564.15 (0.0238675)
```
The RMSE difference is 0.023, which is 2.3%, which is fairly close. With only 5 or 6 bits per channel, we shouldn't expect much better.

But we can improve this (if we are using HDRI) by adding 0.5 to the channels:
```
%IMG7%magick ^
  toes.png ^
  -channel R -evaluate Multiply %%[fx:31/QuantumRange] -evaluate Add 0.5 -evaluate LeftShift 11 ^
  -channel G -evaluate Multiply %%[fx:63/QuantumRange] -evaluate Add 0.5 -evaluate LeftShift 5 ^
  -channel B -evaluate Multiply %%[fx:31/QuantumRange] -evaluate Add 0.5 ^
  +channel ^
  -separate ^
  -evaluate-sequence Add ^
  gray:x.bin

%IMG7%magick compare ^
  -metric RMSE ^
  toes.png ^
  -size 267x233 RGB565:x.bin ^
  NULL:

908.201 (0.0138583)
```

**Reply by crim491** [reply]
#2559 was able to give me what I was looking for. 
Command:
`convert test.bmp -alpha off -define bmp:subtype=RGB565 test-modified.bmp` 
gave me the results I was looking for.

**Reply by t246246** [reply]
I see you want bmp format.  Then `-define bmp:subtype=RGB565` will do it. However it does not do further control how to convert 8bit to 5bit or 6 bit.  So if you need to control it, consider to use snibgo's excellent answer.

**Comment by kdarras** [comment]
Interesting thread... after reading many legacy forum posts which still show up as top results in search engines, I could sadly confirm that `-define bmp:subtype=RGB565`, despite what what it should do and what people say it does, does not do anything useful or what is says: headers stay the same, values for each channel also still max out at 255 after I run `identify -verbose`. The results stays RGB888. Am I correct, is this an issue I should file or is my logic wrong?

Thus, @snibgo 's answer seems most useful but unfortunately, I cannot find out what the syntax should be on Linux (I am new to ImageMagick) in the terminal so I cannot cross-check/confirm.

Running Imagemagick 8:6.9.10.23+dfsg-2.1ubuntu11.4 . However, Software center says this is compiled for a channel bit depth of 16 (not sure if this is relevant).

**Comment by snibgo** [comment]
> Am I correct, is this an issue I should file or is my logic wrong?

You logic is wrong. "identify -verbose" gives metadata about the image _after_ it has been read into memory. If you want metadata about the file structure, IM won't give you that. Instead, use exiftool.

**Comment by kdarras** [comment]
thanks... so I am confused. You are saying that despite the fact that I see:
```
Image: *_565-0.bmp
  Format: BMP (Microsoft Windows bitmap image)
  Class: DirectClass
  Geometry: 591x500+0+0
  Units: PixelsPerCentimeter
  Colorspace: sRGB
  Type: TrueColor
  Base type: Undefined
  Endianess: Undefined
  Depth: 8-bit
  Channel depth:
    red: 8-bit
    green: 8-bit
    blue: 8-bit
  Channel statistics:
    Pixels: 295500
    Red:
      min: 0  (0)
      max: 255 (1)
      mean: 111.501 (0.437261)
      standard deviation: 75.7549 (0.297078)
      kurtosis: -0.98125
      skewness: 0.44847
      entropy: 0.974727
    Green:
      min: 0  (0)
      max: 255 (1)
      mean: 125.324 (0.491467)
      standard deviation: 72.6147 (0.284764)
      kurtosis: -1.07155
      skewness: 0.210208
      entropy: 0.977553
    Blue:
      min: 0  (0)
      max: 255 (1)
      mean: 83.1343 (0.326017)
      standard deviation: 82.948 (0.325286)
      kurtosis: -0.795162
      skewness: 0.78226
      entropy: 0.914463

```

...the image that I created with `-define bmp:subtype=RGB565` actually _is_ a RGB565 BMP?

exiftool tells me it has a bit depth of 16 but the BitsPerSample function does not output anything

Also, when I do `convert *.jpg -define jpg:subtype=RGB565 *_565.jpg` (this is my aim, format-wise), I get a JPEG that exiftool and imagemagick both agree it's RGB888.

In the end, I still wonder how to convert my RGB888 jpegs to RGB565 jpegs. Would you be so kind to provide a Linux syntax for your above code?

**Comment by snibgo** [comment]
> You are saying that despite the fact that I see ... the image that I created with -define bmp:subtype=RGB565 actually is a RGB565 BMP?

I am saying that if you want to know, you should use exiftool.

For example:
```
f:\web\im>%IMG7%magick toes.png -define bmp:subtype=RGB565 x.bmp

f:\web\im>exiftool x.bmp
ExifTool Version Number         : 12.13
File Name                       : x.bmp
Directory                       : .
File Size                       : 122 KiB
File Modification Date/Time     : 2022:01:21 18:39:24+00:00
File Access Date/Time           : 2022:01:21 18:39:24+00:00
File Creation Date/Time         : 2022:01:21 18:39:24+00:00
File Permissions                : rw-rw-rw-
File Type                       : BMP
File Type Extension             : bmp
MIME Type                       : image/bmp
BMP Version                     : Windows V5
Image Width                     : 267
Image Height                    : 233
Planes                          : 1
Bit Depth                       : 16
Compression                     : Bitfields
Image Length                    : 124888
Pixels Per Meter X              : 11811
Pixels Per Meter Y              : 11811
Num Colors                      : Use BitDepth
Num Important Colors            : All
Red Mask                        : 0x0000f800
Green Mask                      : 0x000007e0
Blue Mask                       : 0x0000001f
Alpha Mask                      : 0x00000000
Color Space                     : sRGB
Rendering Intent                : Picture (LCS_GM_IMAGES)
Image Size                      : 267x233
Megapixels                      : 0.062
```
There is one sample of 16 bits per pixel. Note the red, green and blue masks. They have respectively 5, 6, and 5 bits.

> In the end, I still wonder how to convert my RGB888 jpegs to RGB565 jpegs.

I haven't encountered a "RGB565 jpeg" format. Does that exist?

> Would you be so kind to provide a Linux syntax for your above code?

Which above code? This one?
```
magick \
  toes.png \
  -channel R -evaluate Multiply %[fx:31/QuantumRange] -evaluate Add 0.5 -evaluate LeftShift 11 \
  -channel G -evaluate Multiply %[fx:63/QuantumRange] -evaluate Add 0.5 -evaluate LeftShift 5 \
  -channel B -evaluate Multiply %[fx:31/QuantumRange] -evaluate Add 0.5 \
  +channel \
  -separate \
  -evaluate-sequence Add \
  gray:x.bin
```
That creates a headerless binary file, not a JPEG file.

**Comment by kdarras** [comment]
I see, thanks for your explanations... sorry for the confusion.

JPEG RGB565 is a common output of imaging sensors in embedded devices, for instance [((https://www.ovt.com/sensor/ov5645/)](https://www.ovt.com/sensor/ov5645/) - the link may not work though as it is discontinued.

**Comment by snibgo** [comment]
> JPEG RGB565 is a common output ...

Sorry, I can't find "JPEG RGB565" on any page of your link, or anywhere else.

**Reply by fmw42** [reply]
Changelog say:

```
2019-06-03 7.0.8-49 Cristy <quetzlzacatenango@image...>
Add support for RGB565 image format (reference https://imagemagick.org/discourse-server/viewtopic.php?f=2&t=36078).

```
This was probably for BMP, not JPG as in the define

`bmp:subtype=value | BMP channel depth subtypes. The choices are: RGB555, RGB565, ARGB4444, ARGB1555. Only support in BMP (BMP4). BMP3 and BMP2 do not contain header fields to support these options.
`

**Comment by snibgo** [comment]
Yes, there is a BMP 565 format, and a binary (headless) 565 format. See posts in this thread. I have never heard of  a "JPEG 565" format. It may exist. If someone can provide a link, that could be helpful. If the JPEG library supports a "JPEG 565" format then I expect IM could use it.

**Comment by kdarras** [comment]
I dug further into the openMV documentation (the camera I am using) to find out about this (https://docs.openmv.io/library/omv.image.html?highlight=qrcode). While the Omnivision sensor is discontinued and links have expired, I found out in the documentation that the raw RGB565 output from the sensor was converted to a JPEG888. Sorry for my confusion.
